src/pudl/transform/eia176.py

```diff
@@ -0,0 +1,116 @@
+"""Module to perform data cleaning functions on EIA176 data tables."""
+
+import pandas as pd
+from dagster import AssetCheckResult, AssetIn, AssetOut, asset_check, multi_asset
+
+from pudl.logging_helpers import get_logger
+
+logger = get_logger(__name__)
+
+
+@multi_asset(
+    outs={
+        "core_eia176__yearly_company_data": AssetOut(),
+        "core_eia861__yearly_aggregate_data": AssetOut(),
+    },
+)
+def _core_eia176__data(
+    raw_eia176__data: pd.DataFrame,
+) -> tuple[pd.DataFrame, pd.DataFrame]:
+    """Take raw list and return two wide tables with primary keys and one column per variable.
+
+    One table with data for each year and company, one with state- and US-level aggregates per year.
+    """
+    raw_eia176__data["report_year"].astype(int)
+    raw_eia176__data["value"].astype(float)
+    raw_eia176__data["variable_name"] = (
+        raw_eia176__data["line"] + "_" + raw_eia176__data["atype"]
+    )
+
+    long_company = raw_eia176__data.loc[
+        raw_eia176__data.company != "total of all companies"
+    ]
+    aggregate_primary_key = ["report_year", "area"]
+    company_drop_columns = ["itemsort", "item", "atype", "line", "company"]
+
+    wide_company = get_wide_table(
+        long_table=long_company.drop(columns=company_drop_columns),
+        primary_key=aggregate_primary_key + ["id"],
+    )
+
+    long_aggregate = raw_eia176__data.loc[
+        raw_eia176__data.company == "total of all companies"
+    ]
+    wide_aggregate = get_wide_table(
+        long_table=long_aggregate.drop(columns=company_drop_columns + ["id"]),
+        primary_key=aggregate_primary_key,
+    )
+
+    return wide_company, wide_aggregate
+
+
+def get_wide_table(long_table: pd.DataFrame, primary_key: list[str]) -> pd.DataFrame:
+    """Take a 'long' or entity-attribute-value table and return a wide table with one column per attribute/variable."""
+    unstacked = (
+        # we must drop 'id' here and cannot use as primary key because its arbitrary/duplicate in aggregate records
+        # 'id' is a reliable ID only in the context of granular company data
+        long_table.set_index(primary_key + ["variable_name"]).unstack(
+            level="variable_name"
+        )
+    )
+    unstacked.columns = unstacked.columns.droplevel(0).fillna(0)
+    unstacked.columns.name = None  # gets rid of "variable_name" name of columns index
+
+    # TODO instead of "first NA value we see in each column" applied willy-nilly, we could check to see if there are any conflicting non-null values using .count() first.
```

Comment:
I didn't exactly discern what you meant here @jdangerx, if you want to expand. I can probably figure out the intent with a closer look.

Reply:
Ah - yes, I meant `first non-NA value` here, which must have been confusing. Check out this example:

```python
>>> df
   a     b
0  1  10.0
1  1   NaN
2  2  20.0
3  2  21.0
>>> df.groupby("a").first()
      b
a      
1  10.0
2  20.0
```

`.first()` grabs the first non-null value regardless of if there are other non-null values. So for group `a=1` we get the unambiguously right outcome of `b=10`, but for group `a=2`... is `b=20` or `b=21` the right outcome?

If we can tell what the right outcome should be, maybe by sorting by some field, then we should sort by that field before calling `.first()`:

```
>>> df.sort_values("b", ascending=False)
   a     b
3  2  21.0
2  2  20.0
0  1  10.0
1  1   NaN
>>> df.sort_values("b", ascending=False).groupby("a").first()
      b
a      
1  10.0
2  21.0
```

If there's no simple rule, we should check to see that we're not running into this ambiguous behavior by checking that there are no groups with multiple non-NA values before going forward with the `.first()` call:

```python
>>> df.groupby("a").count()
   b
a   
1  1
2  2
>>> (df.groupby("a").count() == 1).all()
b    False
dtype: bool
```

Hopefully that clears things up!

Reply:
We're guaranteed to have one value per group by the time we call `first`, because we're using `unstack`, which doesn't tolerate competing values:

```
>>> df
   company_id variable  value
0           1     var1     11
1           1     var2     12
2           2     var2     22
3           2     var2      0
>>> df.set_index(["company_id", "variable"]).unstack(level="variable")
...
ValueError: Index contains duplicate entries, cannot reshape
```

In fact I don't think we need `first` at all. Seems like a holdover from when we were working with the idea of a sparse matrix.

```
>>> df
   company_id variable  value
0           1     var1     11
1           1     var2     12
2           2     var2     22
>>> df.set_index(["company_id", "variable"]).unstack(level="variable")
           value
variable    var1  var2
company_id
1           11.0  12.0
2            NaN  22.0
```

I'll get the adjustment in my next commit.